### PR TITLE
(POC) Demonstrate extension for iterating TIMESTAMP schema

### PR DESCRIPTION
**Pull request by totten**
Overview
----------------------------------------

(This would contribute toward [https://lab.civicrm.org/dev/core/-/issues/3152's](https://lab.civicrm.org/dev/core/-/issues/3152) approach `#2`. Compare: #23092)

This is an experiment to see what happens if you try to mirror all `TIMESTAMP` columns as `DATETIME[GMT]` columns.

```bash
$ civibuild restore dmaster --no-test --no-cms
$ cv en temporary
$ cv ev 'CRM_Temporary_Schema::setMode("hybrid");' -vv
$ cv api4 Note.getFields +s name,data_type -T
+-------------------+-----------+
| name              | data_type |
+-------------------+-----------+
...
| note_date         | Timestamp |
| created_date      | Timestamp |
| modified_date     | Timestamp |
...
| note_date_gmt     | Timestamp | <== NOTE: New fields in DB. Available to other layers.
| created_date_gmt  | Timestamp | <== NOTE: APIv4 presents DateTime and Timestamp as the same
| modified_date_gmt | Timestamp |
+-------------------+-----------+
```

In theory, this would enable a few phases:

* _Preliminary_: Mainline schema uses `TIMESTAMP`. You can add the `DATETIME[GMT]` columns and start using them for experiments.
* _Transition_: Actively enabled, supporting both `TIMESTAMP` and `DATETIME[GMT]` columns. Incrementally patch apps/UIs/APIs to use new schema.
* _Fade Away_: No longer needed for core purposes. Turn it off by default, but keep it around for backward-compat (with fading support).

Before
--------------------

Core schema is mishmash.

After
--------------------

Core schema is still mishmash.

You can enable the extension `temporary` to play with alternative temporal schema.

Technical Details
--------------------

For example, I did this small check:

* Enabled the extension and switched to `hybrid` mode (as in the "Overview")
* Create a "New Mailing".
* Examine the `civicrm_mailing` table, from the perspective of different time zones. In this log, observe that setting `@@time_zone` causes the `created_date` and `modified_date` to display accordingly; but the `created_date_gmt` and `modified_date_gmt` are always fixed.

```sql
mysql> set @@time_zone = '+0:00'; select id, name, created_date, created_date_gmt, modified_date, modified_date_gmt from civicrm_mailing;
Query OK, 0 rows affected (0.00 sec)

+----+---------+---------------------+---------------------+---------------------+---------------------+
| id | name    | created_date        | created_date_gmt    | modified_date       | modified_date_gmt   |
+----+---------+---------------------+---------------------+---------------------+---------------------+
|  1 | foobar1 | 2022-04-02 09:37:11 | 2022-04-02 09:37:11 | 2022-04-02 09:45:48 | 2022-04-02 09:45:48 |
|  2 | zoop    | 2022-04-02 09:48:24 | 2022-04-02 09:48:24 | 2022-04-02 09:48:39 | 2022-04-02 09:48:39 |
+----+---------+---------------------+---------------------+---------------------+---------------------+
2 rows in set (0.00 sec)

mysql> set @@time_zone = 'America/Chicago'; select id, name, created_date, created_date_gmt, modified_date, modified_date_gmt from civicrm_mailing;
Query OK, 0 rows affected (0.00 sec)

+----+---------+---------------------+---------------------+---------------------+---------------------+
| id | name    | created_date        | created_date_gmt    | modified_date       | modified_date_gmt   |
+----+---------+---------------------+---------------------+---------------------+---------------------+
|  1 | foobar1 | 2022-04-02 04:37:11 | 2022-04-02 09:37:11 | 2022-04-02 04:45:48 | 2022-04-02 09:45:48 |
|  2 | zoop    | 2022-04-02 04:48:24 | 2022-04-02 09:48:24 | 2022-04-02 04:48:39 | 2022-04-02 09:48:39 |
+----+---------+---------------------+---------------------+---------------------+---------------------+
2 rows in set (0.00 sec)

mysql> set @@time_zone = 'America/Denver'; select id, name, created_date, created_date_gmt, modified_date, modified_date_gmt from civicrm_mailing;
Query OK, 0 rows affected (0.01 sec)

+----+---------+---------------------+---------------------+---------------------+---------------------+
| id | name    | created_date        | created_date_gmt    | modified_date       | modified_date_gmt   |
+----+---------+---------------------+---------------------+---------------------+---------------------+
|  1 | foobar1 | 2022-04-02 03:37:11 | 2022-04-02 09:37:11 | 2022-04-02 03:45:48 | 2022-04-02 09:45:48 |
|  2 | zoop    | 2022-04-02 03:48:24 | 2022-04-02 09:48:24 | 2022-04-02 03:48:39 | 2022-04-02 09:48:39 |
+----+---------+---------------------+---------------------+---------------------+---------------------+
2 rows in set (0.00 sec)
```

### CODE REVIEW

File: ext/temporary/CRM/Temporary/Meta.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+use CRM_Temporary_ExtensionUtil as E;
+
+class CRM_Temporary_Options {
+
+  public static function getTimestampModes(): array {
+    return [
+      'auto' => ts('(Auto)'),
+      'hybrid' => ts('Hybrid'),
+      'ts' => ts('TIMESTAMP Only'),
+      'gmt' => ts('DATETIME GMT Only'),
```
Comment (totten) [inline ext/temporary/CRM/Temporary/Meta.php:12]: The description may not have made this clear enough. The above setting is central to what the PR is saying. Basically:

* Take the whole data model and all the TIMESTAMP columns. (Or for smaller scope, set `TEMPORARY_TIMESTAMP_TABLES` and focus on specific tables.)
* For all these tables/fields, you can configure a general mode:
    * `TIMESTAMP Only`: This is what we have today.
    * `Hybrid`: Define both `TIMESTAMP` and `DATETIME-GMT` columns. Mirror data between them.
    * `DATETIME GMT Only`: Get rid of the TIMESTAMP columns

This gives a few knobs that can be dialed-in during a transition-plan.

For example, one transition plan might be "Everything in 9 months":

* __November__: Add extension as optional. Set `TEMPORARY_TIMESTAMP_TABLES` to all tables. Encourage more advanced/engaged folks to enable the extension with `Hybrid` mode. See if any trouble shakes-out.
* __January__: Make extension mandatory. Enable `Hybrid` mode on all sites. But you can still opt-out for a while.
* __February - July__: Old fields are deprecated. Start transitioning all the screens/reports/etc.
* __August__: Start running test-suites in mode `DATETIME GMT Only`. Deprecate `hybrid` and `ts` modes.
* __Next Major Version__: Drop support for `hybrid` and `ts` modes.

I'm not saying that ^^^ is a good or complete plan. There are issues. (Like: On large DBs, upgrades and reconfigurations would create lots of DB work.) A better plan might have some more staging (e.g. table-by-table; or batches of related tables). Or maybe all-tables-at-same-time is good. I'm just trying to illustrate how these are building-blocks for a transition plan.

File: ext/temporary/CRM/Temporary/Schema.php [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+          if (mb_strtoupper($field['default'] ?? '') === 'CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP') {
+            $onUpdate = ["SET NEW.{$gmtField} = UTC_TIMESTAMP();"];
+          }
+          else {
+            $onUpdate = [
+              "IF NEW.{$tsField} <> OLD.{$tsField} THEN",
+              "  SET NEW.{$gmtField} = CONVERT_TZ(NEW.{$tsField}, @@time_zone, '+0:00');",
+              "ELSEIF NEW.{$gmtField} <> OLD.{$gmtField} THEN",
+              "  SET NEW.{$tsField} = CONVER_TZ(NEW.{$gmtField}, '+0:00', @@time_zone);",
```
Comment (demeritcowboy) [inline ext/temporary/CRM/Temporary/Schema.php:57]: CONVER

**Comment by demeritcowboy** [comment]
With the DST changeover coming up this might be a good opportunity to set up some before and after DST tests. Without also changing the setting of the timezone though it might still be off by an hour: https://github.com/civicrm/civicrm-core/blob/f055d6da1ff5520cdf37bc8b3303921f16121697/CRM/Utils/System/Base.php#L727. It always sets it to a numeric offset which doesn't understand DST.

File: ext/temporary/CRM/Temporary/Schema.php (hunk 2 above)
Comment (totten) [reply ext/temporary/CRM/Temporary/Schema.php]: Doh. Confirmed that this typo caused an error. Fixed.

**Comment by totten** [comment]
> (@demeritcowboy) With the DST changeover coming up this might be a good opportunity to set up some before and after DST tests. Without also changing the setting of the timezone though it might still be off by an hour: 

What kind of testing do you have in mind?

Just thinking out loud... IIRC, there were main two problem-scenarios:

1. _Cross Boundary View_: You schedule and send a mail-blast during January (*standard time*).  The timestamp is stored with the true value. In August (*DST*), you come back and read the mail-blast data. The display uses the wrong offset (off by 1 hour).
2. _Cross Boundary Edit_: During January, you schedule a mail-blast for August. The timestamp is stored with the wrong offset (off by 1 hour). Come August, it will send at the wrong time (off by 1 hour).

The theory of `DATETIME[GMT]` approach is that it allows us to fix other screens+integrations (case-by-case):

1. _Cross Boundary View_: If you update a screen to use a different field, then the bug goes away.
    * Ex: "Display `scheduled_date` per MySQL TZ" ==> "Display `scheduled_date_gmt` per PHP TZ"
2. _Cross Boundary Edit_: If you update a screen to use a different field, then the bug goes away.
    * Ex: "Edit `scheduled_date` per MySQL TZ" ==> "Edit `scheduled_date_gmt` per PHP TZ"

This doesn't fix any flawed data produced by CBE's in the past. (AFAICS, that's a major pain in any design. The mitigating factor is that CBE's of `TIMESTAMP` fields  is a minority of edits...)

(Aside: I've assumed that fixes would generally be case-by-case, but for APIv4-based screens, the calls already go through a request-parser... so there may be a way to transparently filter those calls and redirect to the new columns...)

**Comment by totten** [comment]
> (@demeritcowboy) Without also changing the setting of the timezone though it might still be off by an hour: 

Ah, this didn't fully hit me when I reread the other day.

> (@totten) The theory...
> 1. _Cross Boundary View_: If you update a screen to use a different field, then the bug goes away...
> 2. _Cross Boundary Edit_: If you update a screen to use a different field, then the bug goes away... 

So it's not as simple as the theory suggests -- at least, not in `hybrid` mode (*transitional period*).

* Hybrid mode still relies on `@@time_zone`. (The SQL triggers need it.) So it's still faced with the choice between numeric offsets (*buggy but universally portable*) or named timezones (*accurate but not universally portable*).
* Data-access through the new field would be consistent+canonical. But data-access through the old field would exhibit the same old bugs.
* In fact,  if the application has a mix of correct+incorrect behaviors during the transitional period, then the bugs could become *more apparent*.
    * Before, CBE-bugs and CBV-bugs temporarily appear to balance out. The error would only become apparent when your *current time* flips DST.
    * In a transitional period, you have a mix of screens. (Some CBE-buggy; some CBE-ok; some CBV-buggy; some CBV-ok.) You no longer have the appearance where CBE/CBV bugs balance each other.

Still, I don't think there's any other way to accomplish a transition to `DATETIME[GMT]` without a hard break. But maybe there are some things  which make that more manageable:

1. Do the roll-out on a subsystem-by-subsystem basis. (Convert the schema and official UI for `civicrm_mailing` in the same release.)
2. We may not be able to correctly+transparently+affordably rewrite all SQL expressions from MySQL TZ API to PHP TZ API... but deprecation warnings are a different game. They don't need to be perfectly correct. So maybe there's an affordable heuristic for warnings...

**Comment by demeritcowboy** [comment]
I know this is proof-of-concept, but is it intended to be runnable? I'm having some trouble evaluating it because

(a) If I enable gmt mode it crashes almost right away when it does the cache rebuild because `UPDATE civicrm_group SET cache_date = null` fails since there is no cache_date field anymore.
(b) In hybrid mode, I went to create a contact note and it crashed with "created_date_gmt doesn't have a default value". Is this what you mean by buggy during transitional period? If so, I'm iffy on whether a transitional period is worth extra effort?

**Comment by eileenmcnaughton** [comment]
Closing this out as it is more appropriately tracked through gitlab than github as a long-term POC PR

**Pull request closed without merging**